toolchain/check/generic.cpp

```diff
@@ -61,60 +63,48 @@ static auto AddGenericConstantInstToEvalBlock(
 }
 
 namespace {
-// A map from an instruction ID representing a canonical symbolic constant to an
-// instruction within an eval block of the generic that computes the specific
-// value for that constant.
-//
-// We arbitrarily use a small size of 256 bytes for the map.
-// TODO: Determine a better number based on measurements.
-using ConstantsInGenericMap = Map<SemIR::InstId, SemIR::InstId, 256>;
-
 // Substitution callbacks to rebuild a generic constant in the eval block for a
 // generic region.
 class RebuildGenericConstantInEvalBlockCallbacks : public SubstInstCallbacks {
  public:
   // `context` must not be null.
-  RebuildGenericConstantInEvalBlockCallbacks(
-      Context* context, SemIR::GenericId generic_id,
-      SemIR::GenericInstIndex::Region region, SemIR::LocId loc_id,
-      ConstantsInGenericMap& constants_in_generic, bool inside_redeclaration)
-      : context_(context),
-        generic_id_(generic_id),
-        region_(region),
+  RebuildGenericConstantInEvalBlockCallbacks(Context* context,
+                                             SemIR::LocId loc_id)
+      : SubstInstCallbacks(context),
         loc_id_(loc_id),
-        constants_in_generic_(constants_in_generic),
-        inside_redeclaration_(inside_redeclaration) {}
-
-  auto context() const -> Context& { return *context_; }
+        constants_in_generic_(
+            context->generic_region_stack().PeekConstantsInGenericMap()) {}
+
+  auto RebuildType(SemIR::TypeInstId type_inst_id) const
+      -> SemIR::TypeId override {
+    // When building instructions in the eval block, form attached types.
+    return context().types().GetTypeIdForTypeConstantId(
+        context().constant_values().GetAttached(type_inst_id));
+  }
 
   // Check for instructions for which we already have a mapping into the eval
   // block, and substitute them for the instructions in the eval block.
```

Comment:
```suggestion
  // block, and substitute them with the instructions in the eval block.
```
Would this be correct? We look for an inst in the eval block, and replace `inst_id` with the inst from the eval block, right?

The way this reads right now, it sounds like the opposite to me. I noticed this while reading and thinking through your PR comment below.